### PR TITLE
cleaning up array overlap func to avoid temp arrays

### PR DESCRIPTION
**Pull request by srsuryadev**
## Summary

* Optimize the array overlap func not to use temporary arrays, since it slows down some of the queries. 
* Made the array overlap to follow use nested loop algorithm for smaller dataset, so it does not allocate many temporary memory, which is a common case. Then, fall back to the TypedSet based algorithm for larger data size. (Threshold = 200 is chosen based on the performance of both the algorithm for varied sizes, and considered the big int since it is the common case).

## Test plan - 
Unit tests

### Nested Loop (Smaller Data size) + TypedSet (Larger Data Size) (NEW Implementation)
```
Benchmark                              (elementType)  (typeClass)  Mode  Cnt    Score   Error  Units
BenchmarkArraysOverlap.largeBenchmark         BIGINT      ELEMENT  avgt   20   93.777 ± 0.940  ns/op
BenchmarkArraysOverlap.largeBenchmark         BIGINT        ARRAY  avgt   20   50.080 ± 0.609  ns/op
BenchmarkArraysOverlap.largeBenchmark         DOUBLE      ELEMENT  avgt   20   94.467 ± 1.397  ns/op
BenchmarkArraysOverlap.largeBenchmark         DOUBLE        ARRAY  avgt   20  110.302 ± 1.871  ns/op
BenchmarkArraysOverlap.largeBenchmark        VARCHAR      ELEMENT  avgt   20   33.606 ± 0.764  ns/op
BenchmarkArraysOverlap.largeBenchmark        VARCHAR        ARRAY  avgt   20  170.139 ± 8.727  ns/op
BenchmarkArraysOverlap.largeBenchmark        BOOLEAN      ELEMENT  avgt   20    8.531 ± 0.192  ns/op
BenchmarkArraysOverlap.largeBenchmark        BOOLEAN        ARRAY  avgt   20   51.315 ± 2.417  ns/op
BenchmarkArraysOverlap.smallBenchmark         BIGINT      ELEMENT  avgt   20    7.056 ± 0.215  ns/op
BenchmarkArraysOverlap.smallBenchmark         BIGINT        ARRAY  avgt   20   14.845 ± 0.778  ns/op
BenchmarkArraysOverlap.smallBenchmark         DOUBLE      ELEMENT  avgt   20    9.830 ± 0.399  ns/op
BenchmarkArraysOverlap.smallBenchmark         DOUBLE        ARRAY  avgt   20   15.628 ± 0.875  ns/op
BenchmarkArraysOverlap.smallBenchmark        VARCHAR      ELEMENT  avgt   20   12.367 ± 0.154  ns/op
BenchmarkArraysOverlap.smallBenchmark        VARCHAR        ARRAY  avgt   20   28.179 ± 0.533  ns/op
BenchmarkArraysOverlap.smallBenchmark        BOOLEAN      ELEMENT  avgt   20    2.599 ± 0.136  ns/op
BenchmarkArraysOverlap.smallBenchmark        BOOLEAN        ARRAY  avgt   20   15.331 ± 1.044  ns/op
```

### Quicksort + temp arrays (OLD Implementation)
```
Benchmark                              (elementType)  (typeClass)  Mode  Cnt    Score    Error  Units
BenchmarkArraysOverlap.largeBenchmark         BIGINT      ELEMENT  avgt   20   80.889 ±  1.695  ns/op
BenchmarkArraysOverlap.largeBenchmark         BIGINT        ARRAY  avgt   20  171.598 ±  5.257  ns/op
BenchmarkArraysOverlap.largeBenchmark         DOUBLE      ELEMENT  avgt   20  210.416 ±  3.790  ns/op
BenchmarkArraysOverlap.largeBenchmark         DOUBLE        ARRAY  avgt   20  129.309 ±  5.277  ns/op
BenchmarkArraysOverlap.largeBenchmark        VARCHAR      ELEMENT  avgt   20  137.746 ±  1.707  ns/op
BenchmarkArraysOverlap.largeBenchmark        VARCHAR        ARRAY  avgt   20  604.581 ± 26.720  ns/op
BenchmarkArraysOverlap.largeBenchmark        BOOLEAN      ELEMENT  avgt   20   20.214 ±  0.827  ns/op
BenchmarkArraysOverlap.largeBenchmark        BOOLEAN        ARRAY  avgt   20  109.445 ±  5.429  ns/op
BenchmarkArraysOverlap.smallBenchmark         BIGINT      ELEMENT  avgt   20   30.820 ±  0.648  ns/op
BenchmarkArraysOverlap.smallBenchmark         BIGINT        ARRAY  avgt   20  252.607 ± 13.032  ns/op
BenchmarkArraysOverlap.smallBenchmark         DOUBLE      ELEMENT  avgt   20   77.823 ±  3.515  ns/op
BenchmarkArraysOverlap.smallBenchmark         DOUBLE        ARRAY  avgt   20  274.720 ± 10.587  ns/op
BenchmarkArraysOverlap.smallBenchmark        VARCHAR      ELEMENT  avgt   20  233.945 ±  9.512  ns/op
BenchmarkArraysOverlap.smallBenchmark        VARCHAR        ARRAY  avgt   20  302.206 ±  4.770  ns/op
BenchmarkArraysOverlap.smallBenchmark        BOOLEAN      ELEMENT  avgt   20   65.324 ±  2.530  ns/op
BenchmarkArraysOverlap.smallBenchmark        BOOLEAN        ARRAY  avgt   20  248.100 ± 13.110  ns/op
```

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 1]
```diff
@@ -13,202 +13,65 @@
...
-        int leftCurrentPosition = 0;
-        int rightCurrentPosition = 0;
-        while (leftCurrentPosition < leftPositionCount && rightCurrentPosition < rightPositionCount) {
-            if (leftArray.isNull(leftPositions[leftCurrentPosition]) || rightArray.isNull(rightPositions[rightCurrentPosition])) {
-                // Nulls are in the end of the array. Non-null elements do not overlap.
-                return null;
+        boolean hasNull = false;
+        TypedSet typedSet = new TypedSet(type, lookArray.getPositionCount(), "arraysOverlap");
+        for (int i = 0; i < lookArray.getPositionCount(); i++) {
+            if (lookArray.isNull(i)) {
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:56]: You don't need to do this

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 2]
```diff
@@ -13,202 +13,65 @@
...
-        int leftSize = leftArray.getPositionCount();
-        int rightSize = rightArray.getPositionCount();
-
-        if (leftSize == 0 || rightSize == 0) {
-            return false;
-        }
-
-        if (leftLongArray == null || leftLongArray.length < leftSize) {
-            leftLongArray = new long[leftSize * 2];
+            typedSet.add(lookArray, i);
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:60]: Or you can do addNonNull here and keep the nullcheck

**Review by kaikalur — CHANGES_REQUESTED** [review]

**Comment by srsuryadev** [comment]
Microbenchmarks when done locally.

Quicksort + temp arrays

  ```
  arrays_overlap ::   44.108 cpu ms ::    0B peak memory :: in 60.2K,      0B,   1.36M/s,      0B/s :: out 60.2K,   118KB,   1.36M/s,   2.6MB/s
 arrays_overlap ::   43.773 cpu ms ::    0B peak memory :: in 60.2K,      0B,   1.37M/s,      0B/s :: out 60.2K,   118KB,   1.37M/s,  2.62MB/s
```

TypedSet approach

```
   arrays_overlap ::   49.802 cpu ms ::    0B peak memory :: in 60.2K,      0B,   1.21M/s,      0B/s :: out 60.2K,   118KB,   1.21M/s,   2.3MB/s
   arrays_overlap ::   46.780 cpu ms ::    0B peak memory :: in 60.2K,      0B,   1.29M/s,      0B/s :: out 60.2K,   118KB,   1.29M/s,  2.45MB/s
```

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java (hunk 1 above)
Comment (srsuryadev) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:56]: Hi, are you referring to `if (lookArray.isNull(i)) {`, then for the case arrays_overlap(Array [NULL], Array [1, 2]), we need to return null as result right?

Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:56]: So yeah see my next comment use addNoNull so that we can avoid double checking.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java (hunk 2 above)
Comment (srsuryadev) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:56]: Rewrote the logic by adding helper method in TypedSet to check if it containsNullElement().

**Comment by srsuryadev** [comment]
@kaikalur 

* Added a benchmark class similar to BenchmarkArrayDistinct - (BenchmarkArraysOverlap).
* Added containsNullElement helper method() and simplified arrays overlap().

Attached Benchmark results for the arrays overlap 

[arrays_overlap_benchmark_results.txt](https://github.com/prestodb/presto/files/6832042/arrays_overlap_benchmark_results.txt)

Explored the possibility of passing the Block to typedSet constructor for populating, it seems internally TypedSet uses BlockBuilder  as internal representation and they use it as the source for most of the methods and internal logic. I felt converting Block to BlockBuilder may not work well, though the other direction works good. Let me know if you have any suggestions on this.

**Comment by kaikalur** [comment]
> @kaikalur
> 
> * Added a benchmark class similar to BenchmarkArrayDistinct - (BenchmarkArraysOverlap).
> * Added containsNullElement helper method() and simplified arrays overlap().
> 
> Attached Benchmark results for the arrays overlap
> 
> [arrays_overlap_benchmark_results.txt](https://github.com/prestodb/presto/files/6832042/arrays_overlap_benchmark_results.txt)
> 
> Explored the possibility of passing the Block to typedSet constructor for populating, it seems internally TypedSet uses BlockBuilder as internal representation and they use it as the source for most of the methods and internal logic. I felt converting Block to BlockBuilder may not work well, though the other direction works good. Let me know if you have any suggestions on this.

Yeah - we can do that separately.

**Comment by srsuryadev** [comment]
Quicksort + temp arrays

```
Benchmark                         Mode  Cnt   Score   Error  Units
BenchmarkArraysOverlap.benchmark  avgt   60  35.178 ± 1.133  ns/op
```
Nested loop:

```
Benchmark                         Mode  Cnt   Score   Error  Units
BenchmarkArraysOverlap.benchmark  avgt   60  17.091 ± 0.522  ns/op
```

nearly 50% improvement

File: presto-benchmark/src/main/java/com/facebook/presto/benchmark/ArrayFunctionsBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package com.facebook.presto.benchmark;
```
Comment (kaikalur) [inline presto-benchmark/src/main/java/com/facebook/presto/benchmark/ArrayFunctionsBenchmark.java:1]: Maybe we don't need this test anymore. Just remove it. The other benchmark is better.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java [hunk 4]
```diff
@@ -147,6 +147,11 @@ public boolean addNonNull(Block block, int position)
         return false;
     }
 
+    public boolean isContainsNullElement()
+    {
+        return containsNullElement;
+    }
+
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java:154]: Revert this change as well. We don't need it.

**Review by kaikalur — CHANGES_REQUESTED** [review]

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        private PageProcessor pageProcessor;
+
+        @Setup
+        public void setup()
+        {
+            MetadataManager metadata = createTestMetadataManager();
+            FunctionAndTypeManager functionAndTypeManager = metadata.getFunctionAndTypeManager();
+            FunctionHandle functionHandle = functionAndTypeManager.lookupFunction("arrays_overlap", fromTypes(new ArrayType(BIGINT), new ArrayType(BIGINT)));
+
+            List<RowExpression> projections = ImmutableList.of(
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java:94]: Please test all types.

**Review by rongrong — COMMENTED** [review]
We cannot really assume that there won't be arrays of large size. So you probably want to run the benchmark on all types and pick a threshold to switch between this algorithm vs a nlogn algorithm.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java (hunk 5 above)
Comment (srsuryadev) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java:94]: What do you feel will be the threshold to switch nlogn? Do you want to add them as constant in the class itself?

**Comment by kaikalur** [comment]
> We cannot really assume that there won't be arrays of large size. So you probably want to run the benchmark on all types and pick a threshold to switch between this algorithm vs a nlogn algorithm.

I'm not convinced that old approach is good for anything less than a few million entries. But since we have the benchmark, let's add some benchmarks where the sets are completly disjoint (so you get full n^2) and see where it starts breaking. I have feeling it will be a very large number.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java (hunk 4 above)
Comment (srsuryadev) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java:154]: added back this since for O(N log N) used TypedSet as solution

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java (hunk 5 above)
Comment (srsuryadev) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java]: * Tested all types.
* Also used threshold (1000) to switch to O(N log N) solution based on TypedSet. (Used TypedSet since it uses minimal temp arrays. Also, after we optimized the block builder we may get better performance. Also, makes the code much simple and compact.

Approach (Nested loop + Typed Set - O(N log N))

```
Benchmark                              Mode  Cnt    Score   Error  Units
BenchmarkArraysOverlap.largeBenchmark  avgt   20  101.916 ± 3.431  ns/op
BenchmarkArraysOverlap.smallBenchmark  avgt   20   17.060 ± 1.049  ns/op
```

Approach (Quicksort + temp arrays)

```
Benchmark                              Mode  Cnt    Score   Error  Units
BenchmarkArraysOverlap.largeBenchmark  avgt   20   96.646 ± 3.153  ns/op
BenchmarkArraysOverlap.smallBenchmark  avgt   20  119.372 ± 1.524  ns/op
```

Comment (rongrong) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java]: The old implementation is faster on large arrays than on smaller arrays? 😮

Comment (srsuryadev) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java]: It was interesting to me as well. I was not sure, if that's noise in measurements when I was doing in local.

**Comment by rongrong** [comment]
> > We cannot really assume that there won't be arrays of large size. So you probably want to run the benchmark on all types and pick a threshold to switch between this algorithm vs a nlogn algorithm.
> 
> I'm not convinced that old approach is good for anything less than a few million entries. But since we have the benchmark, let's add some benchmarks where the sets are completly disjoint (so you get full n^2) and see where it starts breaking. I have feeling it will be a very large number.

I didn't mean to compare nested loops with the old implementation. I mean to compare nested loops vs typed set to see how many elements is a reasonable threshold to switch between the two.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java (hunk 4 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java:154]: You can just name the function `containsNullElements`

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java (hunk 5 above)
Comment (srsuryadev) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java]: Since a monotonic function with respect to time is expected.

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestTypedSet.java [hunk 6]
```diff
@@ -199,6 +199,23 @@ public void testGetElementPositionRandom()
         testGetElementPositionRandomFor(setWithPassedInBuilder);
     }
 
+    @Test
+    public void testIsContainsNullElement()
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestTypedSet.java:203]: If you update the function name, please also update this as well.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 7]
```diff
@@ -13,202 +13,300 @@
...
+        if (hasNull) {
+            return null;
         }
+        return false;
+    }
 
-        for (int i = 0; i < leftPositionCount; i++) {
-            leftPositions[i] = i;
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = boolean.class)
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:181]: TypeParameterSpecialization for boolean and double is not necessary. `boolean` is only used for boolean type and `double` is only used for `double`. But this is fine if you want consistency. Though there's too much code duplication. It would be nice to just have 5 separate functions for the `equals.invokeExact(xxx)` part and pass that method handle to a single implementation.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Fork(1)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkArraysOverlap
+{
+    private static final int POSITIONS =  10_000;
+    private static final int SMALL_ARRAY_SIZE = 10;
+    private static final int LARGE_ARRAY_SIZE = 1500;
+    private static final int NUM_TYPES = 3;
+    private static final List<Type> TYPES = ImmutableList.of(VARCHAR, BOOLEAN, BIGINT);
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java:76]: Also add double and array or map. You implemented 5 methods, but you are only testing 3 of them here.

Comment (rongrong) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java:76]: Map type configuration as a parameter, so you would test each type in a separate run rather than testing all of them together. You can take a look at `BenchmarkPageProcessor` for examples on how to use `@Param` in benchmarking.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 9]
```diff
@@ -13,202 +13,300 @@
...
+import static com.facebook.presto.common.function.OperatorType.EQUAL;
+import static com.facebook.presto.util.Failures.internalError;
 
 @ScalarFunction("arrays_overlap")
 @Description("Returns true if arrays have common elements")
 public final class ArraysOverlapFunction
 {
-    private int[] leftPositions;
-    private int[] rightPositions;
+    private static final int NESTED_LOOP_THRESHOLD = 1000;
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:38]: How is this value chosen?

Comment (srsuryadev) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:38]: I was doing experiments with 100, 500, 1000, 1500, 2000 (coarse grained level) and compared the performance of nested loop and typed set. Then. chose the value where I could see the performance of Nested loop and typed set are almost similar (i.e the cross-over point).

**Comment by kaikalur** [comment]
Eventually, I think we should just have a clean TypedSet without BlockBuilder and that should work for all sizes as the complexity will be O(m+n).

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java (hunk 5 above)
Comment (kaikalur) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java]: > The old implementation is faster on large arrays than on smaller arrays? 😮

it uses qucksort so not a surprise?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 10]
```diff
@@ -13,202 +13,105 @@
...
-                rightCurrentPosition++;
-            }
-            else if (compareValue < 0) {
-                leftCurrentPosition++;
+                for (int j = 0; j < rightArray.getPositionCount(); j++) {
+                    if (rightArray.isNull(j)) {
+                        hasNull = true;
+                        continue;
+                    }
+                    if (elementType.equalTo(leftArray, i, rightArray, j)) {
```
Comment (srsuryadev) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:64]: Simplified the code duplicity by using equalTo() function in Type, since TypedSet also uses that internally.

Also so slight improvement in the performance for small arrays benchmark.

# Run complete. Total time: 00:03:50

Benchmark                              Mode  Cnt   Score   Error  Units
BenchmarkArraysOverlap.largeBenchmark  avgt   20  96.736 ± 3.395  ns/op
BenchmarkArraysOverlap.smallBenchmark  avgt   20   9.419 ± 0.490  ns/op

**Comment by srsuryadev** [comment]
* Simplified the arrays_overlap function by removing the code duplicity using Type.equals().
* Used @Param for all the combination of types.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java (hunk 9 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:32]: It would be nice to put the benchmark data of that in the PR as well. Thanks!

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 11]
```diff
@@ -13,202 +13,105 @@
...
+                        return true;
+                    }
+                }
             }
-            else {
-                return true;
+            catch (Throwable t) {
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:69]: Is this necessary? I think any uncaught exception eventually will be surfaced as internal error.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java [hunk 12]
```diff
@@ -13,202 +13,105 @@
...
         }
-        return leftArray.isNull(leftPositions[leftPositionCount - 1]) || rightArray.isNull(rightPositions[rightPositionCount - 1]) ? null : false;
+        if (hasNull) {
+            return null;
+        }
+        return false;
     }
 
-    public Boolean genericArraysOverlap(Block leftArray, Block rightArray, AbstractType type)
+    public static Boolean arraysOverlapSetBased(Type type,
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:79]: This should be a private function right? If you make it private, you also don't need to check for 0 position count, or checkArgument that position count is at least `NESTED_LOOP_THRESHOLD`.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java (hunk 11 above)
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java:69]: Do we need this? Does any method in the body throw? If not, remove it.

**Review by kaikalur — CHANGES_REQUESTED** [review]

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+import static com.facebook.presto.sql.analyzer.TypeSignatureProvider.fromTypes;
+import static com.facebook.presto.sql.relational.Expressions.field;
+import static com.facebook.presto.testing.TestingConnectorSession.SESSION;
+
+@SuppressWarnings("MethodMayBeStatic")
+@State(Scope.Thread)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Fork(3)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkArraysOverlap
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArraysOverlap.java:74]: Can you publish the benchmark results in commit message and PR. Thanks!

**Comment by srsuryadev** [comment]
Remeasured the array sizes vs performance for the latest code.

For Array of big ints, iteration_count = 20.

Array Size | Nested Loop (ns/op) | TypedSet Based (ns/op)
-- | -- | --
10 | 7.840 ± 0.934 | 60.939 ± 0.967
100 | 30.530 ± 0.69 | 44.319 ± 0.809
150 | 54.733 ± 0.838 | 59.423 ± 1.066
175 | 64.265 ± 0.855 | 82.797 ± 1.600
200 | 68.491 ± 0.792 | 45.618 ± 3.082
250 | 77.668 ± 4.113 | 51.498 ± 0.681
500 | 138.816 ± 6.987 | 53.045 ± 0.963
1000 | 241.829 ± 5.02 | 56.947 ± 3.510
2000 | 560.196 ± 29.603 | 48.349 ± 0.846


Reconfigured the threshold to 200 after the latest experiment results.

One thing interesting is that performance of the TypedSet has little higher latency for "non 100" array sizes (eg. 10, 150, 175, ...) need to look at this deeper.

Overall Performance:

```
Benchmark                              (elementType)  (typeClass)  Mode  Cnt    Score   Error  Units
BenchmarkArraysOverlap.largeBenchmark         BIGINT      ELEMENT  avgt   20   93.777 ± 0.940  ns/op
BenchmarkArraysOverlap.largeBenchmark         BIGINT        ARRAY  avgt   20   50.080 ± 0.609  ns/op
BenchmarkArraysOverlap.largeBenchmark         DOUBLE      ELEMENT  avgt   20   94.467 ± 1.397  ns/op
BenchmarkArraysOverlap.largeBenchmark         DOUBLE        ARRAY  avgt   20  110.302 ± 1.871  ns/op
BenchmarkArraysOverlap.largeBenchmark        VARCHAR      ELEMENT  avgt   20   33.606 ± 0.764  ns/op
BenchmarkArraysOverlap.largeBenchmark        VARCHAR        ARRAY  avgt   20  170.139 ± 8.727  ns/op
BenchmarkArraysOverlap.largeBenchmark        BOOLEAN      ELEMENT  avgt   20    8.531 ± 0.192  ns/op
BenchmarkArraysOverlap.largeBenchmark        BOOLEAN        ARRAY  avgt   20   51.315 ± 2.417  ns/op
BenchmarkArraysOverlap.smallBenchmark         BIGINT      ELEMENT  avgt   20    7.056 ± 0.215  ns/op
BenchmarkArraysOverlap.smallBenchmark         BIGINT        ARRAY  avgt   20   14.845 ± 0.778  ns/op
BenchmarkArraysOverlap.smallBenchmark         DOUBLE      ELEMENT  avgt   20    9.830 ± 0.399  ns/op
BenchmarkArraysOverlap.smallBenchmark         DOUBLE        ARRAY  avgt   20   15.628 ± 0.875  ns/op
BenchmarkArraysOverlap.smallBenchmark        VARCHAR      ELEMENT  avgt   20   12.367 ± 0.154  ns/op
BenchmarkArraysOverlap.smallBenchmark        VARCHAR        ARRAY  avgt   20   28.179 ± 0.533  ns/op
BenchmarkArraysOverlap.smallBenchmark        BOOLEAN      ELEMENT  avgt   20    2.599 ± 0.136  ns/op
BenchmarkArraysOverlap.smallBenchmark        BOOLEAN        ARRAY  avgt   20   15.331 ± 1.044  ns/op
```

**Review by rongrong — APPROVED** [review]
Thanks for working on this. It's great impact on query performance! Can you please also put the benchmark result in the commit message, that way people can see it in `git log` as well. Thanks!

**Comment by srsuryadev** [comment]
> Thanks for working on this. It's great impact on query performance! Can you please also put the benchmark result in the commit message, that way people can see it in `git log` as well. Thanks!

updated the commit message

**Review by kaikalur — APPROVED** [review]
